### ISSUE TITLE
"/usr/bin/env: python3: No such file or directory" on docker image

### ISSUE THREAD

**Issue by danielvijge** [danielvijge/lms_mixcloud#36]
Trying to play a Mixcloud stream on LMS running on the official docker image gives an error message:
`/usr/bin/env: python3: No such file or directory`

The official docker image does not contain python 3, but python 3.8 is required by the bundled version of the helper application (yt-dlp). Instead of installing python every time the docker image is updated, follow the following steps:

1. Download a [standalone Linux version of yt-dlp](https://github.com/yt-dlp/yt-dlp?tab=readme-ov-file#alternatives). Download the correct version for your architecture (x64, armv7l, aarch64)
2. Save/copy the download to a folder that is mounted inside the docker image, for example the music folder at `/music`, so the binary path becomes `/music/yt-dlp`.
3. Make it executable by running from a terminal `chmod +x /music/yt-dlp` (can be done outside the contain, on the host).
4. In the Mixcloud settings, change the helper application from bundled to system, and in the path text box set `/music/yt-dlp`.

The Mixcloud plugin will now always use this version, and the docker image can be updated without issue.

**Comment by richardhenwood** [comment]
This is great advice for improving the user experience of this excellent plugin! It should be possible to mount the stand-alone `yt-dlp` directly off the host into the container. Something like:
```
podman run -it \
    ...
    -v /local/version/of/yt-dlp:/srv/squeezebox/cache/InstalledPlugins/Plugins/MixCloud/Bin/yt_dlp \
    ...
```
(I haven't tested this myself, as I'm using: `custom-init-sh` to install python3 (and ffmpeg))

**Comment by isavcic** [comment]
If you have the YouTube plugin installed, you can use its yt-dlp at `/config/cache/InstalledPlugins/Plugins/YouTube/Bin/yt-dlp_linux`, just specify it in the Mixcloud plugin settings.

**Comment by tgauchet** [comment]
I am using the youtube yt-dlp, on PCP installed on raspberry pi. I ve set the path to specify it: /mnt/Lyrion2To/slimserver/Cache/InstalledPlugins/Plugins/YouTube/Bin/yt-dlp_linux..
but when I try to play something, I have below errors: 
```
[26-04-25 13:29:19.4801] Plugins::MixCloud::ProtocolHandler::_fetchTrackExtra (250) Failed to determine stream URL for mixcloud://NTSRadio/le-jardin-w-sarah-davachi-3rd-august-2020/
[26-04-25 13:29:19.4806] Plugins::MixCloud::ProtocolHandler::_fetchTrackExtra (251) Tried to execure command: /mnt/Lyrion2To/slimserver/Cache/InstalledPlugins/Plugins/YouTube/Bin/yt-dlp_linux --skip-download --dump-json https://www.mixcloud.com/NTSRadio/le-jardin-w-sarah-davachi-3rd-august-2020/ 2>&1
[26-04-25 13:29:19.4810] Plugins::MixCloud::ProtocolHandler::_fetchTrackExtra (252) /mnt/Lyrion2To/slimserver/Cache/InstalledPlugins/Plugins/YouTube/Bin/yt-dlp_linux: line 0: syntax error: unexpected word (expecting ")")
[26-04-25 13:29:19.4874] Slim::Player::Song::open (426) Error: Couldn't create command line for unk playback for [mixcloud://NTSRadio/le-jardin-w-sarah-davachi-3rd-august-2020/]
```

**Comment by richardhenwood** [comment]
Hi @tgauchet 

I have not encountered your error so cannot help directly. This is a volunteer project, and you may not have a prompt help to debug your problem... but if you are hopeful that someone might be able to help the typical questions you might expect folks to ask you include:
 + please open a new issue for this: this issue is titled '/usr/bin/env python3 not found' - and that is not your issue apprently.
 + on your new issue, please provide versions of the software are you running.
 + on your new issue, please reproduce this error from a reboot and post the complete log output.
 + on your new issue, please report if mixcloud worked in the past and what has changed since then.

As an alternative approach:
I think I typically use the default yt-dlp on PCP. I think that PCP does not include python3, so if my memory serves me correctly, I would manually install python3 (something like `tce-load -wi python3.x.tcz`? or maybe there is something in the GUI that allows one-click install?) and use the defaults for lms_mixcloud. I don't have a PCP instance handy so I cannot be confident of the exact steps, sorry. If I have

**Comment by danielvijge** [comment]
From the error message "line 0: syntax error: unexpected word (expecting ")")" it appears that `Plugins/YouTube/Bin/yt-dlp_linux` isn't a valid script or executable. It might help to open a terminal and see if you can execute it from the directory itself using `./yt-dlp_linux`.

**Comment by tgauchet** [comment]
thank you Daniel & Richard for your answers. 

Firstly my issue is  '/usr/bin/env python3 not found', that is the reason why I am in this post and I have tried the proposed workaround (use the yt-dlp of the  YouTube plugin)...

because of all the wires plugged to the pi board, it is not easy for me to run commands...

Last news is that I have discovered that through PCP extensions, I am able to install  yt-dlp, it has a dependency to python3.11 that is already install. With this method your plugin is working. Problem is that after a reboot,  LMS is not starting anymore (Waiting for valid date before starting LMS..... Done. (2) Starting LMS, this can take some time...Bus error Slimserver is not Running). until I uninstall the yt-dlp... So this issue can be closed. I have a new issue but not with your great plugin.

thanks a lot for your great support.

**Comment by isavcic** [comment]
Don't depend on env. Run an interactive shell in the container, find the python3 binary, use the absolute path.
